function-controller/pkg/controller/autoscaler/delayer.go

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2018-Present the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package autoscaler
+
+import "time"
+
+type Delayer struct {
+	policy        func() time.Duration
+	value         int
+	previousValue int
+	deadline      time.Time
+}
+
+// NewDelayer creates a delayer which delays scaling down to zero until the delay returned by the given policy has elapsed.
+func NewDelayer(delayPolicy func() time.Duration) *Delayer {
+	return &Delayer{
+		policy: delayPolicy,
+	}
+}
+
+func (b *Delayer) Delay(proposal int) *Delayer {
+	if proposal < 0 {
+		panic("Proposed numbers of replicas must be non-negative")
+	}
+
+	now := time.Now()
+	if proposal > 0 {
+		b.value = proposal
+		b.previousValue = proposal
+		b.deadline = now
+	} else if proposal < b.value {
```

Comment:
Isn't that simply `proposal == 0`?

Reply:
No. `proposal == 0` is always true at this point, but we want this condition to be false when `b.value` is zero.

But I take your point about it being a little confusing to refer to `proposal` when we know it's zero. I've rewritten the condition as `b.value > 0` (since proposals are non-negative) to take into account that `proposal == 0`.